README.md
PSL-3604
========


History.txt
PSL-3604

Исправления hardware:

1. В схеме передней панели ошибка - перепутаны выводы
   11 и 12 регистра U4.
 + Исправил на готовой плате перемычками, благо рядом
   находятся нужные переходные отверстия.

2. Процессор на передней панели заменен на STM32F100C8T6B.
   По идее, хватит и STM32F100C6T6B, но не проверял.

3. Напряжение на конденсаторе C20 около 48 В,
   что близко к максимальному (50 В).
 + Заменил конденсатор C20 на 2200 мкФ 63 В.

4. Напряжение на входе стабилизатора -5 В слишком высокое
   (около 20 В).
 + Заменил конденсаторы C1 и C3 на 47 мкФ 100 В.
   Напряжение уменьшилось примерно до 13 В, что нормально.

5. Выброс при переходе между двумя уровнями стабилизации тока.
 + Промоделировал и добавил в схему R98 10 кОм и C51 1 нФ.

6. Ошибка АЦП на несколько единиц.
 + На входе АЦП C2, C4 заменил на 220 пФ.
   Объяснение тут: http://leoniv.livejournal.com/194681.html

7. Делитель COM на плате передней панели желательно сделать
   из прецизионных резисторов с низким ТКС.
 + Поставил прецизионные 1.5 кОм и 49.9 Ом.
 ? Уровень COM нужно уточнять, для этого требуется
   проверить линейность ЦАП на краях диапазона
   с точностью до LSB с включенным буфером.

8. На выходе в состоянии OFF (с выключенным DP) около -0.5 В.
 - Удалить компоненты VD20, R78. R90 увеличить до 1.2 кОм. 

9. Для детектирования состояния Unregulated лучше
   вход компаратора на LM358 отключить от выхода
   усилителя ошибки I и подключить к земле.
 - Проверить работу схемы CV/CC с втекающим выходным током
   от постороннего источника.

10. Фильтры на выходах DAC сделать равнокомпонентными
    47к, 10нФ. Такой фильтр похож на Бесселя,
    а вот фильтр Баттерворта имеет выброс на ПХ,
    что для БП недопустимо.

11. Хорошо бы на входе фильтра ЦАП добавить пассивную RC-цепочку
    (например, 100..680 Ом + 220..10000 пФ)
    для предварительной фильтрации ВЧ-компонетов
    сигнала DitheredDAC.

12. Очень удачным оказалось дизайнерское решение "одно окно",
    LED-индикаторы смотрятся замечательно. Но есть и 
    серьезный недостаток - динамическая индикация создает
    большие пульсации тока потребления платы индикации, что
    является причиной помех на аналоговые узлы источника.
  - Добавить емкость 220 - 1000 мкФ x 6.3 - 16 В на разъем
    питания платы процессора.
  - Разрезать полигон земли платы power таким образом, чтобы
    C24, C25, C26, C27 оказались подключенными к COM,
    а не к AGND. Резистор R49 заменить перемычкой.
    Эта проблема была видна сразу, ну почему я не сделал
    два варианта разделения полигонов с выбором конфигурации
    перемычками...
  - В качестве бредовой идеи - сделать активный стабилизатор
    тока потребления платы индикации :)

13. При выключении питания напряжение спадает плавно,
    на индикации мелькают разные цифры, что некрасиво.
  + Использовал встроенный PVD, котрый при падении напряжения
    питания всё выключает и делает software reset.
  - Проблема осталась, индикация ведет себя при выключении плохо.

14. При выдергивании сетевой вилки при низком выходном напряжении
    на выходе наблюдается выброс до 10 В. Причину пока установить
    не удалось :(

Доработки (для следующей версии):

1. Считывание ошибки Remote Sense.

2. Считывание ошибки супервизора питания. Супервизор должен
   контролировать и основное питание.

3. Мониторинг процессором AC-питания и выключение источника
   при пропуске нескольких полупериодов.

4. Считывание отдельно сигналов CC, CV для индикации состояния
   Unregulated. Оно может появится при просадке сети или
   при большой разнице между Force и Sense.

5. Отключение DownProgrammer отдельным логическим сигналом.

6. Задание тока DownProgrammer, тогда источник можно будет
   использовать в роли электронной нагрузки.

7. Поддержка вентиляторов 3-pin и 4-pin, обработка
   сигнала тахометра.

8. Отказаться от LED-индикаторов в пользу LCD из-за помех?

Список деталей:

+ Радиатор
+ Переходная пластина для транзисторов
+ Прокладки, втулки, винты крепления транзисторов
+ Прижим термометра
+ Стойки платы выходного каскада М2.5х6 =2
+ Стойки платы USB М2.5х? =2
+ Пластина крепления трансформатора
+ Узел крепления трансформатора, винт М6х25 полукругл.
+ Шестигранные стойки М3х35 =6
+ Винты М3х12 потай черные =6
+ Шайбы, гайки М3 =6
+ Винты М3х8 полукругл. =6
+ Передняя панель
+ Стекло дисплея
+ Стойки платы, пластик 12 мм =5
+ Крепление энкодера (2 кольца, пластик)
+ Колпачки кнопок =4
+ Наклейка передней панели
+ Блок клемм, обрезать, вклеить в панель
+ Установить светодиоды, подобрать резисторы
  Резистор для LED_CC уменьшил с 2.2 кОм до 1 кОм
+ Саморезы 2.5 мм =5, привинтить стойки к плате
+ Приклеить стойки к панели, приклеить крепление энкодера
+ Шлейф USB [PW-4] - [PW-4]
+ Шлейф передней панели [IDC-10] - [IDC-10]
+ Шлейф управления выходным каскадом [IDC-10] - [IDC-10]
+ Пара проводов выходной каскад - силовая плата [2x4.8] - [2x4.8]
+ 7 проводов подачи питания [4.8] - [4.8]
+ Шлейф sense [PW-2] - [разделка]
+ Пара проводов к выходным клеммам [4.8] - [разделка]
+ Шлейф управления вентилятором [PW-2] - [PW-2]
+ Удалить желтый провод вентилятора, установить разъем PW-2
+ Сделать переходник [MF-2F] - [FUSE] - [2x4.8]
+ Сделать переходник [2x4.8] - [X2 CAP (22 nF) + MF-2M]
- Сделать резиновые вставки в ножки.

Трансформатор:

  До перемотки: ТТП-104-0,105-УХЛ4,
  I ~220 В (Ixx = 6.7 мА).
  II ~12.45 В (хх) - 68 вит.  (5.46 вит./В).

  Перемотка: 
  Вторичку удалил, намотал две пары обмоток.
  II - (2 x 8.6 В) 2 x 47 вит. в два провода диаметром 1.5 мм (реально намотал 1.16 мм).
  III - (2 х 10.2 В) 2 x 56 вит. (реально намотал 53 вит.) в два провода диаметром 1.16 мм.
  Получились размеры трансформатора D = 93 mm, H = 41 mm.

  Первая пара обмоток заняла неполный слой,
  вторая пара - ровно один слой (поэтому и не домотал 3 витка).
  Вторая пара обмоток, конечно, маловата (а первая - тонковата, было бы самое то 1.5 мм),
  при выходном напряжении 36 В обещанных 100 Вт не получу, а только где-то 70 Вт,

port.h
//----------------------------------------------------------------------------

//Модуль поддержки порта, заголовочный файл

//----------------------------------------------------------------------------

#ifndef PORT_H
#define PORT_H

#include "wakeport.h"

//----------------------------- Константы: -----------------------------------

#define DEVICE_NAME "PSL-3604" //имя устройства
#define BAUD_RATE       19200  //скорость обмена, бод
#define FRAME_SIZE         32  //максимальный размер фрейма, байт

//----------------------------------------------------------------------------
//------------------------------ Класс TPort ---------------------------------
//----------------------------------------------------------------------------

class TPort
{
private:
public:
  TWakePort *WakePort;
  TPort(void);
  void Execute(void);
};

//----------------------------------------------------------------------------
//----------------------------- Коды команд: ---------------------------------
//----------------------------------------------------------------------------

#define CMD_SET_VI 6 //установка напряжения и тока

  //TX: word V, word I, byte S
  //RX: byte Err

  //V = 0..VMAX - напряжение, x0.01 В
  //I = 0..IMAX - ток, x0.001 А
  //S = 0 - выход выключен, 1 - выход включен
  //Err = ERR_NO

#define CMD_GET_VI 7 //чтение установленного напряжения и тока

  //TX:
  //RX: byte Err, word V, word I

  //V = 0..VMAX - напряжение, x0.01 В
  //I = 0..IMAX - ток, x0.001 А
  //Err = ERR_NO

#define CMD_GET_STAT 8 //чтение статуса источника

  //TX:
  //RX: byte Err, byte S

  //S.0 = 1 - выход включен
  //S.1 = 1 - CV
  //S.2 = 1 - CC
  //S.3 = 1 - OVP
  //S.4 = 1 - OCP
  //S.5 = 1 - OPP
  //S.6 = 1 - OTP
  //Err = ERR_NO

#define CMD_GET_VI_AVG 9 //чтение среднего измеренного напряжения и тока

  //TX:
  //RX: byte Err, word VA, word IA

  //VA = 0..VMAX - напряжение, x0.01 В
  //IA = 0..IMAX - ток, x0.001 А
  //Err = ERR_NO

#define CMD_GET_VI_FAST 10 //чтение мгновенного измеренного напряжения и тока

  //TX:
  //RX: byte Err, word VF, word IF

  //VF = 0..VMAX - напряжение, x0.01 В
  //IF = 0..IMAX - ток, x0.001 А
  //Err = ERR_NO

#define CMD_SET_VIP_MAX 11 //установка макс. напряжения, тока и мощности

  //TX: word VM, word IM, word PM
  //RX: byte Err

  //VM = 1000..9999 - напряжение, x0.01 В
  //IM = 1000..9999 - ток, x0.001 А
  //PM = 10..9999 - мощность, x0.1 Вт
  //Err = ERR_NO

#define CMD_GET_VIP_MAX 12 //чтение макс. напряжения, тока и мощности

  //TX:
  //RX: byte Err, word VM, word IM, word PM

  //VM = 1000..9999 - напряжение, x0.01 В
  //IM = 1000..9999 - ток, x0.001 А
  //PM = 10..9999 - мощность, x0.1 Вт
  //Err = ERR_NO

#define CMD_SET_PROT 13 //установка порогов защиты

  //TX: word VP, word IP, word PP
  //RX: byte Err

  //VP = 0..VMAX - напряжение, x0.01 В
  //IP = 0..IMAX - ток, x0.001 А
  //PP = 0..PMAX - мощность, x0.1 Вт
  //Err = ERR_NO

#define CMD_GET_PROT 14 //чтение порогов защиты

  //TX:
  //RX: byte Err, word VP, word IP, word PP

  //VP = 0..VMAX - напряжение, x0.01 В
  //IP = 0..IMAX - ток, x0.001 А
  //PP = 0..PMAX - мощность, x0.1 Вт
  //Err = ERR_NO

#define CMD_SET_PRE 15 //запись пресета

  //TX: byte N, word V, word I
  //RX: byte Err

  //N = 0..9 - номер пресета
  //V = 0..VMAX - напряжение, x0.01 В
  //I = 0..IMAX - ток, x0.001 А
  //Err = ERR_NO, ERR_PA

#define CMD_GET_PRE 16 //чтение пресета

  //TX: byte N
  //RX: byte Err, word V, word I

  //N = 0..9 - номер пресета
  //V = 0..VMAX - напряжение, x0.01 В
  //I = 0..IMAX - ток, x0.001 А
  //Err = ERR_NO, ERR_PA

#define CMD_SET_PAR 17 //установка параметра

  //TX: byte N, word P
  //RX: byte Err

  //N - номер параметра (см. таблицу параметров)
  //P - значение параметра (см. таблицу параметров)
  //Err = ERR_NO, ERR_PA

#define CMD_GET_PAR 18 //чтение параметра

  //TX: byte N
  //RX: byte Err, word P

  //N - номер параметра (см. таблицу параметров)
  //P - значение параметра (см. таблицу параметров)
  //Err = ERR_NO, ERR_PA

#define CMD_GET_FAN 19 //чтение скорости вентилятора и температуры

  //TX:
  //RX: byte Err, byte S, word T

  //S = 0..100 - скорость вентилятора, %
  //T = 0..999 - температура, x0.1°C
  //Err = ERR_NO

#define CMD_SET_DAC 20 //установка кода ЦАП

  //TX: word DACV, word DACI
  //RX: byte Err

  //DACV = 0..65520 - код ЦАП напряжения
  //DACI = 0..65520 - код ЦАП тока
  //Err = ERR_NO

#define CMD_GET_ADC 21 //чтение кода АЦП

  //TX: 
  //RX: byte Err, word ADCV, word ADCI

  //ADCV = 0..65520 - код АЦП напряжения
  //ADCI = 0..65520 - код АЦП тока
  //Err = ERR_NO

#define CMD_SET_CAL 22 //установка калибровочного коэффициента

  //TX: byte N, word K
  //RX: byte Err

  //N - номер коэффициента (см. таблицу коэффициентов)
  //K - значение коэффициента (см. таблицу коэффициентов)
  //Err = ERR_NO, ERR_PA

#define CMD_GET_CAL 23 //чтение калибровочного коэффициента

  //TX: byte N
  //RX: byte Err, word K

  //N - номер коэффициента (см. таблицу коэффициентов)
  //K - значение коэффициента (см. таблицу коэффициентов)
  //Err = ERR_NO, ERR_PA

//----------------------------------------------------------------------------

#endif
